cli/tests/test_simplify_parents_command.rs

```diff
@@ -0,0 +1,142 @@
+
+    (test_env, repo_path)
+}
+
+fn create_commit(test_env: &TestEnvironment, repo_path: &Path, name: &str, parents: &[&str]) {
+    let mut args = vec!["new", "-m", name];
+    args.extend(parents);
+    test_env.jj_cmd_ok(repo_path, &args);
+
+    std::fs::write(repo_path.join(name), format!("{name}\n")).unwrap();
+    test_env.jj_cmd_ok(repo_path, &["bookmark", "create", name]);
+}
+
+#[test]
+fn test_simplify_parents_no_args() {
+    let (test_env, repo_path) = create_repo();
+
+    let stderr = test_env.jj_cmd_cli_error(&repo_path, &["simplify-parents"]);
+    insta::assert_snapshot!(stderr, @r###"
+    error: the following required arguments were not provided:
+      <--sources <SOURCES>|--revisions <REVISIONS>>
+
+    Usage: jj simplify-parents <--sources <SOURCES>|--revisions <REVISIONS>>
+
+    For more information, try '--help'.
+    "###);
+}
+
+#[test]
+fn test_simplify_parents_no_commits() {
+    let (test_env, repo_path) = create_repo();
+
+    let (stdout, stderr) =
+        test_env.jj_cmd_ok(&repo_path, &["simplify-parents", "-r", "root() ~ root()"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @r###"
+    Nothing changed.
+    "###);
+}
+
+#[test]
+fn test_simplify_parents_immutable() {
+    let (test_env, repo_path) = create_repo();
+
+    let stderr = test_env.jj_cmd_failure(&repo_path, &["simplify-parents", "-r", "root()"]);
+    insta::assert_snapshot!(stderr, @r###"
+    Error: The root commit 000000000000 is immutable
+    "###);
+}
+
+#[test]
+fn test_simplify_parents_no_change() {
+    let (test_env, repo_path) = create_repo();
+
+    create_commit(&test_env, &repo_path, "a", &["root()"]);
+    create_commit(&test_env, &repo_path, "b", &["a"]);
+    let stdout = test_env.jj_cmd_success(&repo_path, &["log", "-r", "all()", "-T", "description"]);
+    insta::assert_snapshot!(stdout, @r###"
+    @  b
+    ○  a
+    ◆
+    "###);
+
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["simplify-parents", "-s", "@-"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @r###"
+    Nothing changed.
+    "###);
+
+    let stdout = test_env.jj_cmd_success(&repo_path, &["log", "-r", "all()", "-T", "description"]);
+    insta::assert_snapshot!(stdout, @r###"
+    @  b
+    ○  a
+    ◆
+    "###);
+}
+
+#[test_case(&["simplify-parents", "-r", "@", "-r", "@-"] ; "revisions")]
+#[test_case(&["simplify-parents", "-s", "@-"] ; "sources")]
+fn test_simplify_parents_redundant_parent(args: &[&str]) {
+    let (test_env, repo_path) = create_repo();
+
+    create_commit(&test_env, &repo_path, "a", &["root()"]);
+    create_commit(&test_env, &repo_path, "b", &["a"]);
+    create_commit(&test_env, &repo_path, "c", &["a", "b"]);
+    let stdout = test_env.jj_cmd_success(&repo_path, &["log", "-r", "all()", "-T", "description"]);
+    insta::allow_duplicates! {
+    insta::assert_snapshot!(stdout, @r###"
+    @    c
+    ├─╮
+    │ ○  b
+    ├─╯
+    ○  a
+    ◆
+    "###);
```

Comment:
Another interesting case to me (that I think we should test) is `jj simplify-parents -r 'all()'` for something like

```
    @    d
    ├─╮
    ○ │  c
    │ ○  b
    ├─╯
    ○  a
    ◆
```

I think it should be a no-op, and I haven't read the code carefully enough to check whether it will be.

Reply:
Thanks for adding! I hadn't noticed when you did it. I'll hopefully take one last look at this soon.